bee-network/src/conns/manager.rs

```diff
@@ -0,0 +1,245 @@
+    internal_event_sender: InternalEventSender,
+}
+
+impl ConnectionManagerConfig {
+    pub fn new(
+        local_keys: identity::Keypair,
+        // TODO: allow multiple bind addresses
+        bind_address: Multiaddr,
+        peers: PeerList,
+        banned_addrs: BannedAddrList,
+        banned_peers: BannedPeerList,
+        internal_event_sender: InternalEventSender,
+    ) -> Result<Self, Error> {
+        // Create underlying Tcp connection and negotiate Noise and Mplex/Yamux
+        let transport = build_transport(&local_keys).map_err(|_| Error::CreatingTransportFailed)?;
+
+        let mut peer_listener = transport
+            .listen_on(bind_address.clone())
+            .map_err(|_| Error::BindingAddressFailed(bind_address))?;
+
+        let listen_address =
+            if let Some(Some(Ok(ListenerEvent::NewAddress(listen_address)))) = peer_listener.next().now_or_never() {
+                trace!("listening address = {}", listen_address);
+                listen_address
+            } else {
+                return Err(Error::NotListeningError);
+            };
+
+        trace!("Accepting connections on {}.", listen_address);
+
+        LISTEN_ADDRESSES.write().unwrap().push(listen_address);
+
+        Ok(Self {
+            peers,
+            banned_peers,
+            banned_addrs,
+            peer_listener,
+            internal_event_sender,
+        })
+    }
+}
+
+#[async_trait]
+impl<N: Node> Worker<N> for ConnectionManager {
+    type Config = ConnectionManagerConfig;
+    type Error = Infallible;
+
+    fn dependencies() -> &'static [TypeId] {
+        vec![TypeId::of::<PeerManager>()].leak()
+    }
+
+    async fn start(node: &mut N, config: Self::Config) -> Result<Self, Self::Error> {
+        let ConnectionManagerConfig {
+            peers,
+            banned_peers,
+            banned_addrs,
+            peer_listener,
+            internal_event_sender,
+            ..
+        } = config;
+
+        node.spawn::<Self, _, _>(|shutdown| async move {
+            info!("Listener started.");
+
+            let mut incoming = ShutdownStream::new(shutdown, peer_listener);
+
+            while let Some(Ok(listener_event)) = incoming.next().await {
+                //
+                if let Some((upgrade, peer_address)) = listener_event.into_upgrade() {
+                    // TODO: try again to move this block into its own function (beware: lifetime issues ahead!!!)
+
+                    // Prevent accepting from banned addresses.
+                    let peer_address_str = peer_address.to_string();
+                    if banned_addrs.contains(&peer_address_str) {
+                        trace!("Ignoring peer. Cause: {} is banned.", peer_address_str);
+                        NUM_LISTENER_EVENT_PROCESSING_ERRORS.fetch_add(1, Ordering::Relaxed);
+                        continue;
+                    }
+
+                    let (peer_id, muxer) = match upgrade.await {
+                        Ok(u) => u,
```

Comment:
maybe rename "u" -> "peer"?

Reply:
the `u` stands for `upgrade` which we are awaiting. Are you also fine with me renaming it to `upgrade`? I think `peer` would be not so fitting, because we not only receive a peer (or to more exact its id), but also get a back a *multiplexer*.

Reply:
Yes that sounds good to me!